### ISSUE TITLE
SDMolSupplier should support streams

### ISSUE THREAD

**Issue by ribeaud** [rdkit/rdkit#1938]
Currently and up to my understanding, you can use `SDMolSupplier` the following ways:
- `Chem.SDMolSupplier('file_name.sdf')` (providing a file name)
- ```supp = Chem.SDMolSupplier();supp.SetData(open('file_name.sdf').read())``` (providing a content)

This has been confirmed by following [thread](https://sourceforge.net/p/rdkit/mailman/message/27914348/) as well. This [link](https://github.com/rdkit/rdkit-orig/blob/master/rdkit/Chem/Suppliers/UnitTestSDMolSupplier.py) points to a test file where `SDMolSupplier` seems to support streaming but I was NOT able to get it working.

Because a **SDF** file could be located in a _zip_ file, on a _ftp_ server or in the _cloud_, `SDMolSupplier` should support streaming or some kind of filesystem abstraction library like [PyFilesystem](https://www.pyfilesystem.org/).

Kind regards,

christian

**Comment by ribeaud** [comment]
If this does NOT request any **C/C++** skills, I could have a look at it, assuming you give me some hints how this should be implemented.

**Comment by bp-kelley** [comment]
This is a touch confusing.  The SDMolSupplier provides access to a random access stream.  For the streams of which you are interested, you'll need a ForwardSDMolSupplier.  Here is a socket client example (note that in python you need to call socket.makefile to make it file like)

```import socket
from rdkit.Chem import ForwardSDMolSupplier

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
host ="127.0.0.1"
port =8000
s.connect((host,port))
s.makefile()

def get(str):
   s.send('e'.encode())
   for mol in ForwardSDMolSupplier(s):
       print(mol.GetNumAtoms())


while 1:
   get(s)

s.close ()
```

**Comment by greglandrum** [comment]
To elaborate on Brian's example, here's a snippet showing how you can read from a `.sdf.gz` file:
```
In [7]: with gzip.open('./mols.1000.sdf.gz') as gzf:
   ...:     suppl = Chem.ForwardSDMolSupplier(gzf)
   ...:     ms = [x for x in suppl if x is not None]
   ...:
```

**Comment by greglandrum** [comment]
No, it's not currently possible to use `GetItemText()` with the `ForwardSDMolSupplier`. The class just reads from the stream and does not cache the values read.

**Comment by UnixJunkie** [comment]
it would make my life easier if both Chem.SDMolSupplier and Chem.SDWriter could both accept .sdf or .sdf.gz files...
